casper/src/main/scala/coop/rchain/casper/SafetyOracle.scala

```diff
@@ -132,111 +131,66 @@ sealed abstract class SafetyOracleInstances {
+              val creatorJustificationOrGenesis = block.justifications
+                .find(_.validator == block.sender)
+                .fold(block.blockHash)(_.latestBlockHash)
+              DagOperations
+                .bfTraverseF[F, BlockHash](List(latestByValidatorHash)) { blockHash =>
+                  ProtoUtil.getCreatorJustificationAsListUntilGoalInMemory(
+                    blockDag,
+                    blockHash,
+                    validator,
+                    b => b == creatorJustificationOrGenesis
+                  )
+                }
+                .pure[F]
+            case None => StreamT.empty[F, BlockHash].pure[F]
+          }
 
         def neverEventuallySeeDisagreement(
             first: Validator,
             second: Validator
         ): F[Boolean] =
-          for {
-            maybeFirstLatest <- blockDag.latestMessage(first)
-            result <- maybeFirstLatest match {
-                       case Some(firstLatestBlock) =>
-                         val justificationHashes =
-                           firstLatestBlock.justifications.map(_.latestBlockHash)
-                         for {
-                           justificationBlockSecondList <- justificationHashes.flatTraverse {
-                                                            justificationHash =>
-                                                              blockDag
-                                                                .lookup(justificationHash)
-                                                                .map(
-                                                                  _.filter(_.sender == second).toList
-                                                                )
-                                                          }
-                           justificationBlockSecond = justificationBlockSecondList.head
-                           _ = assert(
-                             justificationBlockSecondList
-                               .forall(b => b.blockHash == justificationBlockSecond.blockHash)
-                           )
-                           potentialDisagreements <- filterChildren(
-                                                      justificationBlockSecond,
-                                                      second
-                                                    )
-                           result <- potentialDisagreements.forallM { potentialDisagreement =>
-                                      computeCompatibility(
-                                        blockDag,
-                                        candidateBlockHash,
-                                        potentialDisagreement
-                                      )
-                                    }
-                         } yield result
-                       case None => false.pure[F]
-                     }
-          } yield result
+          blockDag.latestMessage(first).flatMap {
```

Comment:
Could be simplified in a similar manner as `getCreatorJustificationAsListUntilGoalInMemory`